src/cards/polymorph.ts

```diff
@@ -0,0 +1,194 @@
+            }
+          }
+        }
+      }
+
+      return state;
+    },
+  },
+};
+
+function polymorphUnit(fromUnit: Unit.IUnit, underworld: Underworld, prediction: boolean, toUnitId?: string): Unit.IUnit | undefined {
+  // If a specific ID isn't passed in, choose a random one
+  if (!toUnitId) {
+    let possibleUnitTypes = Object.values(allUnits).filter(u => isModActive(u, underworld) && u.id != fromUnit.unitSourceId);
+    if (Unit.isBoss(fromUnit.unitSourceId)) {
+      // Get all boss units
+      possibleUnitTypes = possibleUnitTypes
+        .filter(u => Unit.isBoss(u.id));
+    } else {
+      // Get all units with budget and spawn chance
+      possibleUnitTypes = possibleUnitTypes
+        .filter(u => u.spawnParams && u.spawnParams.probability > 0 && u.spawnParams.budgetCost);
+    }
+
+    // We have to seed this to prevent multiplayer desync
+    // ID is required here to ensure that every polymorph changes the seed to prevent looping
+    const seed = seedrandom(`${getUniqueSeedString(underworld)} - ${fromUnit.id}`);
+    toUnitId = chooseObjectWithProbability(possibleUnitTypes.map((p, index) => {
+      // Units are weighted by their difference in budget.
+      // Units twice as far away in the budget are half as likely to be chosen.
+      const budgetDiff = Math.abs((allUnits[fromUnit.unitSourceId]?.spawnParams?.budgetCost || 0) - (p.spawnParams?.budgetCost || 0));
+      return { unitSource: p, probability: (budgetDiff == 0) ? 1 : 1 / budgetDiff }
+    }), seed)?.unitSource.id;
+
+    if (toUnitId == undefined) {
+      console.error("Polymorph failed to choose a new unit type.");
+      return undefined;
+    }
+  }
+  const toSourceUnit = allUnits[toUnitId];
+  if (!toSourceUnit) {
+    console.error('Unit with id', toUnitId, 'does not exist. Have you registered it in src/units/index.ts?');
+    return undefined;
+  }
+
+  // Cases for polymorphing AI / Player
+  if (fromUnit.unitType != UnitType.PLAYER_CONTROLLED) {
+    let unit: Unit.IUnit = Unit.create(
+      toSourceUnit.id,
+      fromUnit.x,
+      fromUnit.y,
+      fromUnit.faction,
+      toSourceUnit.info.image,
+      UnitType.AI,
+      toSourceUnit.info.subtype,
+      { ...toSourceUnit.unitProps, isMiniboss: fromUnit.isMiniboss, originalLife: fromUnit.originalLife },
+      underworld,
+      prediction
+    );
+
+    if (unit != undefined) {
+      // Keep Modifiers
+      for (const modifierKey of Object.keys(unit.modifiers)) {
+        const modifier = allModifiers[modifierKey];
+        const modifierInstance = unit.modifiers[modifierKey];
+        if (modifier && modifierInstance) {
+          if (modifier?.add) {
+            modifier.add(unit, underworld, prediction, modifierInstance.quantity, modifierInstance);
+          }
+        } else {
+          console.error("Modifier doesn't exist? This shouldn't happen.");
+        }
+      }
+
+      // Cleanup old unit and remove it from targets
+      Unit.cleanup(fromUnit, false);
```

Comment:
This line doesn't remove the unit from targets, that is handled in the calling function right now.  Let's pass the effectState into this function and handle the cleanup and adding the new unit to the target array here